syntax = "proto3";

package CoreScript;

// The main service for interacting with the CoreScript engine in Revit.
service CoreScriptRunner {
  rpc ExecuteScript(ExecuteScriptRequest) returns (ExecuteScriptResponse);
  rpc GetStatus(GetStatusRequest) returns (GetStatusResponse);
  rpc GetScriptMetadata(GetScriptMetadataRequest) returns (GetScriptMetadataResponse);
  rpc GetScriptParameters(GetScriptParametersRequest) returns (GetScriptParametersResponse);
  rpc GetCombinedScript(GetCombinedScriptRequest) returns (GetCombinedScriptResponse);
  rpc CreateAndOpenWorkspace(CreateWorkspaceRequest) returns (CreateWorkspaceResponse);
  // New RPC for fetching the entire script manifest
  rpc GetScriptManifest(GetScriptManifestRequest) returns (ScriptManifestResponse);
}

// Request for the GetScriptManifest RPC
message GetScriptManifestRequest {
  string agent_scripts_path = 1;
}

// General purpose empty message
message Empty {}

message CreateWorkspaceRequest {
  string script_path = 1;
  string script_type = 2;
}

message CreateWorkspaceResponse {
  string workspace_path = 1;
  string error_message = 2;
}

message ScriptFile {
  string file_name = 1;
  string content = 2;
}

message ExecuteScriptRequest {
  string script_content = 1;
  bytes parameters_json = 2;
  string source = 3;
}

message StructuredOutputItem {
  string type = 1;
  string data = 2; 
}

message OutputSummary {
  string type = 1; // e.g., "string", "table", "console", "return_value", "error"
  string message = 2; // General summary message

  TableSummary table = 3;
  ConsoleSummary console = 4;
  ReturnValueSummary return_value_summary = 5;
}

message TableSummary {
  int32 row_count = 1;
  repeated string column_headers = 2;
  repeated string truncated_rows_json = 3; // JSON serialized strings of rows
}

message ConsoleSummary {
  int32 line_count = 1;
  repeated string truncated_lines = 2; // First 5 lines
}

message ReturnValueSummary {
  string type = 1;
  string value = 2; // String representation of the return value
}

message ExecuteScriptResponse {
  bool is_success = 1;
  string output = 2;
  string error_message = 3;
  repeated string error_details = 4;
  repeated StructuredOutputItem structured_output = 5;
  OutputSummary output_summary = 6; // New field for structured summary
}

message GetStatusRequest {
}

message GetStatusResponse {
  bool rserver_connected = 1;
  bool revit_open = 2;
  string revit_version = 3;
  bool document_open = 4;
  string document_title = 5;
  string document_type = 6;
}

message GetScriptMetadataRequest {
  repeated ScriptFile script_files = 1;
}

message GetScriptMetadataResponse {
  ScriptMetadata metadata = 1;
  string error_message = 2;
}

message GetScriptParametersRequest {
  repeated ScriptFile script_files = 1;
}

// Represents the metadata for a single script (single-file or multi-file).
// This is the new, finalized format.
message ScriptMetadata {
  string name = 1;                  // Name of the script (filename or folder name)
  string file_path = 2;             // Relative path from the agent workspace root
  string script_type = 3;           // "single-file" or "multi-file"
  string description = 4;
  string author = 5;
  repeated string categories = 6;
  repeated string dependencies = 7;
  string document_type = 8;
  repeated string usage_examples = 9;        // New field for usage examples
  string website = 10; 
  string last_run = 11;
}

// The response containing the full manifest of all available scripts.
message ScriptManifestResponse {
  repeated ScriptMetadata scripts = 1;
  string error_message = 2; // Field for communicating errors
}
message GetScriptParametersResponse {
  repeated ScriptParameter parameters = 1;
  string error_message = 2;
}

message ScriptParameter {
  string name = 1;
  string type = 2;
  string default_value_json = 3;
  string description = 4;
  repeated string options = 5;
}

message GetCombinedScriptRequest {
  repeated ScriptFile script_files = 1;
}

message GetCombinedScriptResponse {
  string combined_script = 1;
  string error_message = 2;
}